### PR TITLE
Fixed #25816 -- Made SpatialRefSysMixin properties work correctly if GDAL is unavailable.

### PR DESCRIPTION
**Pull request by sir-sigurd**
https://code.djangoproject.com/ticket/25816

### CODE REVIEW

File: django/contrib/gis/db/backends/base/models.py [hunk 1]
```diff
@@ -62,9 +62,14 @@ def ellipsoid(self):
...
             return self.srs.ellipsoid
         else:
-            m = self.spheroid_regex.match(self.wkt)
+            wkt = self.wkt.replace('\n', '')
+            m = self.spheroid_regex.match(wkt)
             if m:
-                return (float(m.group('major')), float(m.group('flattening')))
+                major = float(m.group('major'))
+                flattening = float(m.group('flattening'))
+                # This is how semiminor calculated by GDAL.
```
Comment (timgraham) [inline django/contrib/gis/db/backends/base/models.py]: `# GDAL's semiminor calculation.`

File: django/contrib/gis/db/backends/base/models.py [hunk 2]
```diff
@@ -84,12 +89,16 @@ def datum(self):
...
+        return self.wkt.startswith('COMPD_CS')
+
+    @property
     def projected(self):
         "Is this Spatial Reference projected?"
         if gdal.HAS_GDAL:
             return self.srs.projected
         else:
-            return self.wkt.startswith('PROJCS')
+            return self.wkt.startswith('PROJCS') or self.compound and 'PROJCS' in self.wkt
```
Comment (timgraham) [inline django/contrib/gis/db/backends/base/models.py]: You might use parenthesis to be explicit about the groupings and improve readability in these longer expressions.

File: django/contrib/gis/db/backends/base/models.py [hunk 3]
```diff
@@ -123,33 +132,38 @@ def linear_units(self):
...
-            return None
+        elif self.geographic and not self.compound:
+            return 1.0
         else:
             m = self.units_regex.match(self.wkt)
-            return m.group('unit')
+            return float(m.group('unit'))
+
+    def _get_geogcs_units_match(self):
+        # We need only units declared under GEOGCS.
```
Comment (timgraham) [inline django/contrib/gis/db/backends/base/models.py]: Extract units declared ...

File: django/contrib/gis/db/backends/base/models.py [hunk 4]
```diff
@@ -123,33 +132,38 @@ def linear_units(self):
...
         "Returns the angular units."
         if gdal.HAS_GDAL:
             return self.srs.angular_units
-        elif self.projected:
-            return None
         else:
-            m = self.units_regex.match(self.wkt)
-            return m.group('unit')
+            m = self._get_geogcs_units_match()
+            DEGREE_IN_RAD = 0.0174532925199433 # constant used by GDAL
```
Comment (timgraham) [inline django/contrib/gis/db/backends/base/models.py]: two spaces before #

File: tests/gis_tests/test_spatialrefsys.py [hunk 5]
```diff
@@ -143,3 +143,36 @@ def test_add_entry(self):
...
+            'projected',
+            'local',
+            'geographic',
+            'linear_name',
+            'linear_units',
+            'angular_name',
+            'angular_units',
+        ]
+        for db_srs in SpatialRefSys.objects.all():
+            # Make sure that we have valid wkt.
```
Comment (timgraham) [inline tests/gis_tests/test_spatialrefsys.py]: Maybe `# Skip invalid wkts` (+explain how that happens?)

File: tests/gis_tests/test_spatialrefsys.py [hunk 6]
```diff
@@ -143,3 +143,36 @@ def test_add_entry(self):
...
+            try:
+                SpatialReference(db_srs.wkt)
+            except Exception:
+                continue
+
+            for attr in attrs:
+                try:
+                    gdal_val = getattr(db_srs, attr)
+                except GDALException:
```
Comment (timgraham) [inline tests/gis_tests/test_spatialrefsys.py]: explain when the exception happens?

**Comment by claudep** [comment]
With each Django version, I wonder if it would be the right time to make GDAL a required dependency for `contrib.gis`. Then we wouldn't need this code any longer (and in many other places as well).
@sir-sigurd any opinion on that?

**Comment by sir-sigurd** [comment]
@claudep, [here](https://code.djangoproject.com/ticket/5433)'s initial ticket to make GDAL optional.
Motivation was:

> The GDAL library is magnitudes more complicated to install than GEOS and is not required for the basic GIS functionality (WKT, geometry operations, querying). And GDAL won't work on windows anytime soon.

I don't think that it's complicated to install GDAL and it seems that it's available on windows, so I think we could do it.

**Comment by claudep** [comment]
GDAL used to be complicated to install on some platforms.
Thanks for your opinion, I will still raise that on the geodjango/django-developers mailing lists before going further.

**Comment by timgraham** [comment]
Closing since GDAL will become a required dependency (#6777).

**Pull request closed without merging**